models/synthetic_data/synthetic_data_for_pipeline_transform/generate.py

```diff
@@ -0,0 +1,287 @@
+
+    rnd_pts = np.random.rand(8) * (10 * noise)
+    pts2 = np.float32([[0 + rnd_pts[0], 0 + rnd_pts[1]],
+                       [photo_w - rnd_pts[2], 0 + +rnd_pts[3]],
+                       [photo_w - rnd_pts[4], photo_h - rnd_pts[5]],
+                       [0 + rnd_pts[6], photo_h - rnd_pts[7]]
+                       ])
+
+    matrix = cv2.getPerspectiveTransform(pts1, pts2)
+    cv2.warpPerspective(img, matrix, dsize=(photo_w, photo_h), dst=photo_img, borderMode=cv2.BORDER_TRANSPARENT)
+
+    writing_pts = img_info["writing_area"]["verts_xy"]
+
+    out_writing_pts = []
+    for pt in writing_pts:
+        src = np.array([[[pt[0], pt[1]]]]).astype("float32")
+        out_writing_pts.append((cv2.perspectiveTransform(src, matrix) - 1).reshape(2))
+
+    return photo_img, out_writing_pts
+
+
+def create_one_and_save_it(noise: float = 1, set_name: str = ""):
+    """
+    function to create a useful sample and save it to the local disk, given an amount of noise and a name for the
+    set that it is part of
+    """
+    if set_name != "":
+        set_name = set_name + "\\"
+
+    file_name_base = str(uuid.uuid4())
+    file_name_img_X_input = os.path.dirname(__file__) + "\\data\\" + set_name + file_name_base + ".X_input.png"
+    file_name_data = os.path.dirname(__file__) + "\\data\\" + set_name + file_name_base + ".json"
+    file_name_img_y_label = os.path.dirname(__file__) + "\\data\\" + set_name + file_name_base + ".y_label.png"
+
+    os.makedirs(os.path.dirname(file_name_img_X_input), exist_ok=True)
+
+    img, img_info = create_fake_paper_with_writing(5, 25, noise=noise)
+    img2, pts = randomize_photo_transform(img, img_info, photo_sizes[0], noise=noise)
+    img3 = cv2.resize(img2, input_size, interpolation=cv2.INTER_AREA)
+    scaleX = input_size[0] / photo_sizes[0][1]
+    scaleY = input_size[1] / photo_sizes[0][0]
+
+    pts = np.array(pts)
+    pts[:, 0] = pts[:, 0] * scaleX
+    pts[:, 1] = pts[:, 1] * scaleY
+    img3 = cv2.cvtColor(img3, cv2.COLOR_RGB2BGR)
+
+    cv2.imwrite(file_name_img_X_input, img3)
+    img_hash = hashlib.md5(open(file_name_img_X_input, 'rb').read()).hexdigest()
+
+    img_y_label = cv2.resize(img, input_size, interpolation=cv2.INTER_AREA)
+    cv2.imwrite(file_name_img_y_label, cv2.cvtColor(img_y_label, cv2.COLOR_RGB2BGR))
+
+    data = {"y_label_points": pts.tolist(),
+            "y_label_image_file": path.basename(file_name_img_y_label),
+            "X_input_image_file": path.basename(file_name_img_X_input),
+            "X_input_file_hash": img_hash
+            }
+
+    with open(file_name_data, 'w') as f:
+        json.dump(data, f)
+
+
+def create_bunch(how_many: int, noise: float = 1, cores: int = 4, noise_start: int = -1, noise_end: int = -1,
+                 noise_step: int = -1):
+    """function to create many synthetic samples, given a noise level and how many to create."""
+    # creates a list of processes waiting to be started and completed.
+    if noise_start == -1:
```

Comment:
the majority of this function's logic with processes could and should be replaced with a [multiprocessing pool](https://stackoverflow.com/questions/20887555/dead-simple-example-of-using-multiprocessing-queue-pool-and-locking). Managing processes manually is not recommended because there a lot of gotchas when it comes to manually handling process objects.

Reply:
This one was of those moments where I decided I was tired of googling things and that it would be easier to do it by hand.. Really should have just looked it up (again). You know how it is when you look up something for the 10th time! 

I'll get to it soon.